### PR TITLE
Straight distance2d expression

### PR DESCRIPTION
**Pull request by Koyaani**
## Description

This PR adds the straight_distance2d function in the expressions.

Hello,
I'm Antoine Facchini and I joined Oslandia at the end of June, I discover the world of GIS. I have almost finished my studies, I'm doing my last internship. During my studies, I worked in different computer fields, now I'm a C++ / Python developer. So, here is my 1st PR, my 1st participation ^^

All tips are welcome !

### CODE REVIEW

File: resources/function_help/json/straight_distance_2d [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "straight_distance_2d",
+  "type": "function",
+  "groups": ["GeometryGroup"],
+  "description": "Returns the straight distance of a geometry. The geometry must be a curve (circularstring, linestring).",
```
Comment (DelazJ) [inline resources/function_help/json/straight_distance_2d:5]:
```suggestion
  "description": "Returns the direct/euclidean distance between the first and last vertex of a geometry. The geometry must be a curve (circularstring, linestring).",
```
I think we better explain what this straightdistance represents. Picked at https://qgis.org/api/3.20/classQgsCurve.html#ac40baa1bf589ed744e7aa56ac3580f63

**Comment by DelazJ** [comment]
Welcome aboard @Koyaani

File: src/core/expression/qgsexpressionfunction.cpp [hunk 2]
```diff
@@ -3585,6 +3585,21 @@ static QVariant fcnSinuosity( const QVariantList &values, const QgsExpressionCon
   return QVariant( curve->sinuosity() );
 }
 
+static QVariant fcnStraightDistance2d( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
```
Comment (troopa81) [inline src/core/expression/qgsexpressionfunction.cpp:3590]:
```suggestion
  const QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
```

**Review by troopa81 — COMMENTED** [review]
Welcome @Koyaani

File: src/core/expression/qgsexpressionfunction.cpp [hunk 3]
```diff
@@ -3585,6 +3585,21 @@ static QVariant fcnSinuosity( const QVariantList &values, const QgsExpressionCon
   return QVariant( curve->sinuosity() );
 }
 
+static QVariant fcnStraightDistance2d( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  const QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
+  const QgsCurve *curve = qgsgeometry_cast< const QgsCurve * >( geom.constGet() );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpressionfunction.cpp:3591]:
```suggestion
  const QgsCurve *curve = geom.constGet() ? qgsgeometry_cast< const QgsCurve * >( geom.constGet()->simplifiedTypeRef() ) : nullptr;
```

There's a special situation which needs handling here, which is that the input geometry is a MultiLineString geometry containing only a single part. This happens a lot, e.g. if the input layer is a shapefile then all geometries will be reported as multilinestrings, even if they only have a single line part. So it's something we need to also allow.

Calling QgsAbstractGeometry::simplifiedTypeRef() is the easiest/fastest way to handle this. simplifiedTypeRef will return either the original geometry if it's already a single-line type, OR the first linestring part if it's a multilinestring containing a single part, OR the original geometry ref if it's neither of these (i.e. a multilinestring with multiple parts). So ultimately it's a really quick way of handling the "single line part in a multilinestring" case without involving any copying of geometries!

File: tests/src/core/testqgsexpression.cpp [hunk 4]
```diff
@@ -1325,6 +1325,14 @@ class TestQgsExpression: public QObject
...
+      QTest::newRow( "straight_distance_2d not geom" ) << "straight_distance_2d('g')" << true << QVariant();
+      QTest::newRow( "straight_distance_2d null" ) << "straight_distance_2d(NULL)" << false << QVariant();
+      QTest::newRow( "straight_distance_2d point" ) << "straight_distance_2d(geom_from_wkt('POINT(1 2)'))" << true << QVariant();
+      QTest::newRow( "straight_distance_2d multi linestring" ) << "straight_distance_2d(geom_from_wkt('MULTILINESTRING( (0 0, 1 1), (2 2, 3 3) )'))" << true << QVariant();
+      QTest::newRow( "straight_distance_2d linestring" ) << "straight_distance_2d(geom_from_wkt('LINESTRING(1 0, 4 4)'))" << false << QVariant( 5.0 );
+      QTest::newRow( "straight_distance_2d linestring" ) << "round(straight_distance_2d(geom_from_wkt('LINESTRING(1 4, 3 5, 5 0)')), 3)" << false << QVariant( 5.657 );
+      QTest::newRow( "straight_distance_2d closed linestring" ) << "straight_distance_2d(geom_from_wkt('LINESTRING(2 2, 3 6, 2 2)'))" << false << QVariant( 0.0 );
+      QTest::newRow( "straight_distance_2d circularstring" ) << "round(straight_distance_2d(geom_from_wkt('CircularString (20 30, 50 30, 10 50)')), 3)" << false << QVariant( 22.361 );
```
Comment (nyalldawson) [inline tests/src/core/testqgsexpression.cpp:1335]: Would be good to also add a test for a multilinestring with a single part (following the comment above)

**Comment by nyalldawson** [comment]
Nice work! Welcome aboard!! :partying_face:

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**